### DISCUSSION TITLE
How to use Supabase with asyncpg?

### DISCUSSION THREAD

**Discussion by nsbradford** [supabase/supabase#20775] [Questions]
For reference:
- https://github.com/MagicStack/asyncpg/issues/1058
- https://github.com/MagicStack/asyncpg/issues/837
- https://github.com/sqlalchemy/sqlalchemy/issues/6467
 
TLDR: [asyncpg](https://github.com/MagicStack/asyncpg) is go-to Postgres client, but it has incompatibility with pgbouncer because of issues with prepared statements. Supabase uses it's own in-house [Supavisor](https://supabase.com/blog/supavisor-1-million), but still seems to have the same issue. 

There are workarounds if you have access to the pgbouncer instance, but there doesn't seem to be a way to configure Supavisor easily? Are there any other workarounds for this?

Workaround details: https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#prepared-statement-name
> To prevent a buildup of useless prepared statements in your application, it’s important to use the [NullPool](https://docs.sqlalchemy.org/en/20/core/pooling.html#sqlalchemy.pool.NullPool) pool class, and to configure PgBouncer to use [DISCARD](https://www.postgresql.org/docs/current/sql-discard.html) when returning connections. The DISCARD command is used to release resources held by the db connection, including prepared statements. Without proper setup, prepared statements can accumulate quickly and cause performance issues.

**Comment by awalias** [comment]
Hey @nsbradford 

Have created an issue for this one on the supavisor repo here: https://github.com/supabase/supavisor/issues/287

**Reply by nsbradford** [reply]
thanks!
